Add gsf module

lispandfound opened the pull request:
@claudio525 Mentioned that the parsing code in ucgmsim/visualization#91 may have broader application and so should be included in qcore. I agree, so this pull request contains the GSF parsing code.

lispandfound commented:
I have also added a new coordinate transformation module, and all the GSF related code from the type5 branch of Pre-processing. The coordinate transformation module is required for the GSF meshgrid generator. @claudio525 I know you approved the trivial parsing code, but it now includes quite a bit more. I will also add a description of the GSF file format to the wiki because I don't think it's documented anywhere.

lispandfound commented on qcore/gsf.py:
```diff
@@ -0,0 +1,210 @@
+    dips: np.ndarray,
+    rakes: np.ndarray,
+    resolution: int,
+):
```
I don't like this function the way that it is currently. I have a `fault` module in type5 atm that encapsulates lengths, widths, strikes, dips, etc into a Fault object rather than this parallel array nonsense. Perhaps I should include this fault module into qcore? I think it has some merit as the geometry code I wrote there could be useful for other projects (like for example, the NSHMDB repo).

lispandfound replied:
If I did that, I'd also do the fault module properly, with multiple ways of instantiating faults (from corners, from centroid + strike, dip, etc). This is because the NSHMDB can read fault corners from the database, but EMOD3D specifies faults with centroids, but fundamentally they represent the same object.

joelridden replied:
Not against having a shared Fault object.

We have in the past created a classdef file which contained this https://github.com/ucgmsim/Pre-processing/blob/f86c185c4a87ba13b3a413e78132e8ca22fa6fb3/VM/models/classdef.py#L4 (Used to also be one in empirical engine till it was removed for over complicating the problem)

Did not handle things very well but could be good to maybe bring these all together under 1 hood if implemented properly and isn't too much "overhead" just to create a fault object to get results from functions.

claudio525 replied:
This would line up with the `preprocessing` -> `source_model` repo change that we will hopefully do in the near future.

joelridden commented on qcore/coordinates.py:
```diff
@@ -0,0 +1,62 @@
+    np.ndarray
+        An array of shape (N, 3) containing NZTM coordinates (x, y, depth).
+    """
+    nztm_coords = np.array(
+        _WGS2NZTM.transform(wgs_depth_coordinates[:, 0], wgs_depth_coordinates[:, 1]),
+    ).T
+    return np.append(nztm_coords, wgs_depth_coordinates[:, 2].reshape((-1, 1)), axis=-1)
```
This can be simplified to just
`nztm_coords = np.array(
        _WGS2NZTM.transform(wgs_depth_coordinates[:, 0], wgs_depth_coordinates[:, 1], wgs_depth_coordinates[:, 2])).T`

joelridden replied:
These comments would be nice to address?

joelridden commented on qcore/coordinates.py:
```diff
@@ -0,0 +1,62 @@
+    np.ndarray
+        An array of shape (N, 3) containing WGS84 coordinates (latitude, longitude, depth).
+    """
+    wgs_coords = np.array(
+        _NZTM2WGS.transform(nztm_coordinates[:, 0], nztm_coordinates[:, 1]),
+    ).T
+    return np.append(wgs_coords, nztm_coordinates[:, 2].reshape((-1, 1)), axis=-1)
```
Same thing as other function

joelridden commented on qcore/gsf.py:
```diff
@@ -0,0 +1,210 @@
+    """
+    with open(gsf_filepath, mode="r", encoding="utf-8") as gsf_file_handle:
+        # we could use pd.read_csv with the skiprows argument, but it's not
+        # as versatile as simply skipping the first n rows with '#'
```
I assume thats because n rows is an unknown length which is different per file?

lispandfound replied:
Yeah the number of comment lines prior to the number of points line is not defined. I generate GSF files with just one comment line, rob uses 3 lines, both are acceptable.

joelridden commented on qcore/gsf.py:
```diff
@@ -0,0 +1,210 @@
+
+    Parameters
+    ----------
+    gsf_file_handle : TextIO
```
actually a str input?

lispandfound replied:
Ah yeah good catch

sungeunbae reviewed (commented):
Happy in general - just a few comments

lispandfound commented on qcore/gsf.py:
```diff
@@ -0,0 +1,210 @@
+
+
+def write_fault_to_gsf_file(
+    gsf_file_handle: TextIO,
```
Yeah that should be a Path, you're right

lispandfound commented:
@sungeunbae @joelridden I have addressed these concerns. I'll leave the unhappy parallel array gsf writing function for now but include a fault class into qcore that I can use in NSHMDB, Pre-Processing and any places later on where the database or type-5 code gets incorporated.

joelridden reviewed (commented):
Just think the other 2 comments for the coordinates.py functions would be good to simplify

claudio525 commented on qcore/gsf.py:
```diff
@@ -0,0 +1,211 @@
+    np.ndarray
+        The meshgrid of the rectangular planar region.
+    """
+    origin = coordinates.wgs_depth_to_nztm(origin)
```
I think it might be worth breaking this up into some blocks, and then adding some block comments. There is a fair few steps and not all of them are trivial imo.

claudio525 commented on qcore/gsf.py:
```diff
@@ -0,0 +1,211 @@
+            dip_step = width / gridpoint_count_in_length(width * 1000, resolution)
+            for point in meshgrid:
+                gsf_file_handle.write(
+                    f"{point[1]:11.5f} {point[0]:11.5f} {point[2] / 1000:11.5e} {strike_step:11.5e} {dip_step:11.5e} {strike:6.1f} {dip:6.1f} {rake:6.1f} {-1.0:8.2f} {-1.0:8.2f} {i:3d}\n"
```
This line is possibly a bit long? Black doesn't break up string lines, so has to be done manually.

claudio525 reviewed (commented):
Couple of minor last minute comments. Otherwise looks great to me.

joelridden commented on qcore/coordinates.py:
```diff
@@ -36,6 +36,8 @@ def wgs_depth_to_nztm(wgs_depth_coordinates: np.ndarray) -> np.ndarray:
     np.ndarray
         An array of shape (N, 3) containing NZTM coordinates (x, y, depth).
     """
+    if wgs_depth_coordinates.shape == (3,):
```
Based on the docstring I would assume this is always the case? Maybe docstring needs to change

lispandfound replied:
No this is to allow code like the following: 

```wgs_depth_to_nztm(np.array([lat, lon, depth]))```

Otherwise you have to do an ugly

```wgs_depth_to_nztm(np.array([lat, lon, depth]).reshape((1, -1)))```

lispandfound replied:
I can update the docstring, however.

joelridden commented on qcore/coordinates.py:
```diff
@@ -59,8 +61,10 @@ def nztm_to_wgs_depth(nztm_coordinates: np.ndarray) -> np.ndarray:
     np.ndarray
         An array of shape (N, 3) containing WGS84 coordinates (latitude, longitude, depth).
     """
+    if nztm_coordinates.shape == (3,):
```
Same comment as above

joelridden commented on qcore/gsf.py:
```diff
@@ -109,55 +111,58 @@ def coordinate_meshgrid(
-                gsf_file_handle.write(
-                    f"{point[1]:11.5f} {point[0]:11.5f} {point[2] / 1000:11.5e} {strike_step:11.5e} {dip_step:11.5e} {strike:6.1f} {dip:6.1f} {rake:6.1f} {-1.0:8.2f} {-1.0:8.2f} {i:3d}\n"
-                )
+        for j in range(n_dip):
```
Not sure how I feel about the variable j. Maybe dip_index or depth_index? As it's the index as you go down the dip / go deeper. Maybe even i could change to plane_index

sungeunbae commented on qcore/gsf.py:
```diff
@@ -0,0 +1,216 @@
+    """
+    with open(gsf_filepath, mode="r", encoding="utf-8") as gsf_file_handle:
+        # we could use pd.read_csv with the skiprows argument, but it's not
+        # as versatile as simply skipping the first n rows with '#'
```
Interesting

claudio525 reviewed (approved):
Nice work!
